dftimewolf/lib/collectors/gce_disk_copy.py

```diff
@@ -0,0 +1,229 @@
+          copying disks.
+    """
+    if not (remote_instance_names or disk_names):
+      self.ModuleError(
+          'You need to specify at least an instance name or disks to copy',
+          critical=True)
+
+    if stop_instances and not remote_instance_names:
+      self.ModuleError(
+          'You need to specify an instance name to stop the instance',
+          critical=True)
+
+    self.source_project = gcp_project.GoogleCloudProject(
+        source_project_name, default_zone=destination_zone)
+    if destination_project_name:
+      self.destination_project = gcp_project.GoogleCloudProject(
+          destination_project_name, default_zone=destination_zone)
+    else:
+      self.destination_project = self.source_project
+
+    self.remote_instance_names = (
+        remote_instance_names.split(',') if remote_instance_names else [])
+    self.disk_names = disk_names.split(',') if disk_names else []
+    self.all_disks = all_disks
+    self.stop_instances = stop_instances
+
+  def PreProcess(self) -> None:
+    """Organise any disks to be copied.
+
+    Process uses GCEDisk containers, so we create those containers and store
+    them in the state.
+    """
+    try:
+      # Disks from the csv list passed in
+      for d in self.disk_names:
+        self.state.StoreContainer(containers.GCEDisk(d))
+
+      # Disks from the instances passed in
+      for i in self.remote_instance_names:
+        try:
+          for d in self._GetDisksFromInstance(i, self.all_disks):
+            self.state.StoreContainer(containers.GCEDisk(d))
+
+        except lcf_errors.ResourceNotFoundError:
+          self.ModuleError(
+            message=f'Instance "{i}" not found or insufficient permissions',
+            critical=True)
+    except HttpError as exception:
+      if exception.resp.status == 403:
+        self.ModuleError(
+            '403 response. Do you have appropriate permissions on the project?',
+            critical=True)
+      if exception.resp.status == 404:
+        self.ModuleError(
+            'GCP resource not found. Maybe a typo in the project / instance / '
+            'disk name?',
+            critical=True)
+      self.ModuleError(str(exception), critical=True)
+
+  def Process(self, container: containers.GCEDisk) -> None:
+    """Copies a disk to the destination project.
+
+    Args:
+      container: GCEDisk container referencing the disk to copy.
+    """
+    self.logger.info(f'Disk copy of {container.name} started...')
+
+    try:
+      new_disk = gcp_forensics.CreateDiskCopy(
+          self.source_project.project_id,
+          self.destination_project.project_id,
+          self.destination_project.default_zone,
+          disk_name=container.name)
+    except lcf_errors.ResourceNotFoundError as exception:
+      self.logger.error(f'Could not find disk "{container.name}": {exception}')
+      self.warned = True
+      self.ModuleError(str(exception), critical=True)
```

Comment:
yeah I wasn't sure either way. I've reorganised it so it will throw a critical error if all disks fail, but will continue on if at least one succeeds.

Reply:
Great! Thanks.